### ISSUE TITLE
Semantics of span_iterator

### ISSUE THREAD

**Issue by MikeGitb** [Microsoft/GSL#437]
I'd like to put up the semantics of `span_iterator` for discussion. 

**Currently**:
They very much behave, as if `span` was a classic container and they would point to some elements inside a span object.

1. They are invalidated as soon as the corresponding span goes out of scope or is modified (assigned a new value), even if the original range from which the span was constructed is unchanged
2. You can't compare iterators from two spans, even if those spans are copies of each other.

In particular, as we are encouraged to pass spans by value, this makes passing around iterators a bit dangerous:

1. The following function signature is illegal, because the returned iterator refers to the local (and now ouot of scope) copy of the span (see my comments on #427):

       gsl::span<int>::iterator myFind(gsl::span<int>range, int);

2. Even if we fix above function signature to pass the span by reference, things like the following would still be illegal:

       auto iter1 = myFind(mySpan.subspan(0,10),5);
       // ... use iter1
       // ERROR: iter1 became invalid after above's semicolon 

       auto iter2 = myFind(gsl::span<int>(my_vector),5);
       // ... use iter2
       // ERROR: iter2 became invalid after above's semicolon 

3. Also the fact that the following won't work might be a suprise for some users:

		struct Foo {
			Foo() { ... }
			gsl::span<int> elements() { return _data; }
		private:
			std::array<int, 10> _data;
		};

		Foo e;
		std::sort(e.elements().begin(), e.elements().end()); // begin and end belong to different spans
          
I'm not saying, they are particularly common scenarious or that there isn't a different way to do things in the first place, but they represent pitfalls when working with gsl::span that seem unfortunate to me.

**Alternative**:
Personally I'd prefer it if `span_iterator` would behave more like a range checked pointer into the original array, in which all the aboves examples would be well defined. AFAIK that is also how `string_view` iterators will behave in c++17 (please correct me if I'm wrong).

Implementation wise that would require that 
 - An iterator should contain an index and a copy of the span from which it was constructed (not a pointer to the original span). 
 - Comparing two iterators would be the same as comparing the memory addresses they refer to.

**Comment by neilmacintosh** [comment]
@MikeGitb I think having a `gsl::span::iterator` with a lifetime that is valid and independent of the `gsl::span` that created it is not the right design. It encourages another level of distance between the lifetime of an actual underlying sequence, and the things that point into it. That usually makes it more difficult to get lifetime semantics correct, not less difficult.

I think each of the examples you have highlighted above are good cases where the code is wrong in a more subtle way. The problems of 1. and 2. can be achieved if you replace `gsl::span` with `std::vector`, and the problem around 3. is not unique to `gsl::span` either.

If you are accepting a parameter by-value and then returning an iterator derived from that parameter, it should smell like a design error (because most likely it is one). If you receive a return of an iterator from something passed by-reference, then you would want to be very careful about temporary lifetimes - again whether the type is `gsl::span` or a standard container.

Comparing two iterators from different instances of `std::map` or `std::vector` or `std::string` does not yield anything good, and I do not think we want to start encouraging this with `gsl::span`. `std::string_view` is its own animal, and I respect it as such, but I do not know if I agree with every design decision that went into it.

So in summary, none of these examples suggest to me that the lifetime of `gsl::span::iterator` should differ from that of `gsl::span`. I'm unaware of any significant scenario or pattern that we either block or make painfully difficult with the current design.

**Comment by fuchsto** [comment]
Also, if I may add, the range concept defines (and explains):

> Range iterators cannot outlive their ranges

<http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4128.html#range-iterators-cannot-outlive-their-ranges>

**Comment by MikeGitb** [comment]
> It encourages another level of distance between the lifetime of an actual underlying sequence, and the things that point into it.

Why? Currently there are two levels: An iterator refers to a span and a span refers to a sequence. With the proposed changes there would only be one: The iterator directly refers to the sequence.

> The problems of 1. and 2. can be achieved if you replace gsl::span with std::vector, and the problem around 3. is not unique to gsl::span either.

The analogy with vector is imho a bit misleading, as span is semantically not equivalent to `std::vector` but `std::vector&`.

Imho `foo(gsl::span<int> range)`  is promoted as a replacement for foo(int* data, size_t size) and `foo(std::vector<int>& container)` style functions, both of which would allow the return of a pointer or iterator respectively.

> Comparing two iterators from different instances of std::map or std::vector or std::string does not yield anything good, and I do not think we want to start encouraging this with `gsl::span`.

Again, the equivalent of `gsl::span<int> elements()` in my example would not be `std::array<int, 10> elements()`, but `std::array<int, 10>& elements()`. for which the example would be fine.

> I'm unaware of any significant scenario or pattern that we either block or make painfully difficult with the current design.

Me neither. Just makes some things imho more inconvenient and/or dangerous. But OK. Now that I've been bitten by it once, I hope it will not happen to me again.

**Comment by aurelienrb** [comment]
Hello :)

As explained in #461, I've also be bitten by 3., i.e. replacing:

```cpp
const std::vector<int> & get() const;
```
with:
```cpp
span<const int> get() const;
```

It is very likely that someone will do something like `auto it = x.get().begin();` and get a crash, while it works fine when returning a `const vector<int> &`.

I don't see why this restriction makes `span` usage safer? For me it violated the principle of least astonishment.

> Comparing two iterators from different instances of std::map or std::vector or std::string does not yield anything good, and I do not think we want to start encouraging this with gsl::span. 

This can be enforced with an implementation like that (see the `assert` in `operator==`):

```cpp
template<typename T>
class iterator {
public:
    iterator(T & begin, size_t length) : ptr(&begin), size(length) {
    }
    iterator() = default;
    iterator(const iterator &) = default;
    iterator & operator=(const iterator & other) = default;

    T * operator->() const {
        return ptr;
    }

    T & operator*() const {
        return *ptr;
    }

    iterator & operator++() {
        assert(size > 0);
        ++ptr;
        --size;
        if (size == 0) {
            ptr = nullptr;
        }
        return *this;
    }

    bool operator==(const iterator & other) const {
        assert(ptr == nullptr || other.ptr == nullptr ||
            &ptr[size] == &other.ptr[other.size]);  // <-- check we have same span iterators
        return size == other.size;
    }

    bool operator!=(const iterator & other) const {
        return !(*this == other);
    }

private:
    T * ptr = nullptr;
    size_t size = 0;
};
```
